e2e/vegeta_test.go

```diff
@@ -39,75 +38,179 @@ const (
 
 // ===== Vegeta Helper Functions =====
 
-// createRPCTarget
-// • Returns a vegeta.Targeter for the specified RPC method
-func createRPCTarget(
-	gatewayURL string,
-	jsonrpcReq jsonrpc.Request,
-	headers http.Header,
-) vegeta.Targeter {
-	return func(tgt *vegeta.Target) error {
-		body, err := json.Marshal(jsonrpcReq)
-		if err != nil {
-			return err
+// runServiceTest runs the E2E test for a single EVM service in a test case.
+func runServiceTest(
+	t *testing.T,
+	ctx context.Context,
+	targets []vegeta.Target,
+	serviceConfig ServiceConfig,
+	serviceType serviceType,
+	methodCount int,
+	summary *serviceSummary,
+) (serviceTestFailed bool) {
+	results := make(map[string]*methodMetrics)
+	var resultsMutex sync.Mutex
+
+	// Get methods to test from the summary
+	methods := summary.methodsToTest
+
+	// Create a map for progress bars and summary calculation (with the same config for all methods)
+	methodConfigMap := make(map[string]ServiceConfig)
+	// Create a map to associate methods with their targets
+	methodTargets := make(map[string]vegeta.Target)
+
+	for i, method := range methods {
+		methodConfigMap[method] = serviceConfig
+		if i < len(targets) {
+			methodTargets[method] = targets[i]
 		}
-		// Set up the HTTP POST target
-		tgt.Method = http.MethodPost
-		tgt.URL = gatewayURL
-		tgt.Body = body
-		tgt.Header = headers
-		return nil
 	}
+
+	progBars, err := newProgressBars(methods, methodConfigMap)
+	if err != nil {
+		t.Fatalf("Failed to create progress bars: %v", err)
+	}
+	defer func() {
+		if err := progBars.finish(); err != nil {
+			fmt.Printf("Error stopping progress bars: %v", err)
+		}
+	}()
+
+	var methodWg sync.WaitGroup
+	for _, method := range methods {
+		methodWg.Add(1)
+
+		target, exists := methodTargets[method]
+		if !exists {
+			fmt.Printf("Warning: No target found for method %s, skipping...\n", method)
+			methodWg.Done()
+			continue
+		}
+
+		go func(ctx context.Context, method string, serviceConfig ServiceConfig, target vegeta.Target) {
+			defer methodWg.Done()
+
+			metrics := runMethodAttack(
+				ctx,
+				method,
+				serviceConfig,
+				serviceType,
+				methodCount,
+				target,
+				progBars.get(method),
+			)
+
+			resultsMutex.Lock()
+			results[method] = metrics
+			resultsMutex.Unlock()
+
+		}(ctx, method, serviceConfig, target)
+	}
+	methodWg.Wait()
+
+	if err := progBars.finish(); err != nil {
+		fmt.Printf("Error stopping progress bars: %v", err)
+	}
+
+	// We use the same methodConfigMap we created earlier for the summary calculation
+
+	calculateServiceSummary(t, methodConfigMap, results, summary, &serviceTestFailed)
+	return serviceTestFailed
+}
+
+// runMethodAttack executes the attack for a single JSON-RPC method and returns metrics.
+func runMethodAttack(
+	ctx context.Context,
```

Comment:
Meta point: not a blocker but at least a TODO:

This file needs more encapsulation overall: multiple functions are well aware of basically all aspects of the functionality, i.e. tests.
We need to split and encapsulate the functionality so each function has a minimal set of inputs, responsibilities to track.

Reply:
I've updated these methods to just take a `TestService` struct.